nativelib/src/main/resources/gc/commix/Marker.c

```diff
@@ -0,0 +1,367 @@
+        if (Heap_IsWordInHeap(heap, field)) {
+            ObjectMeta *fieldMeta = Bytemap_Get(bytemap, field);
+            if (ObjectMeta_IsAllocated(fieldMeta)) {
+                Marker_markObject(heap, stats, outHolder, bytemap,
+                                  (Object *)field, fieldMeta);
+            }
+        }
+    }
+}
+
+void Marker_markPacket(Heap *heap, Stats *stats, GreyPacket* in, GreyPacket **outHolder) {
+    Bytemap *bytemap = heap->bytemap;
+    if (*outHolder == NULL) {
+        GreyPacket *fresh = Marker_takeEmptyPacket(heap, stats);
+        assert(fresh != NULL);
+        *outHolder = fresh;
+    }
+    while (!GreyPacket_IsEmpty(in)) {
+        Object *object = GreyPacket_Pop(in);
+
+        if (Object_IsArray(object)) {
+            if (object->rtti->rt.id == __object_array_id) {
+                ArrayHeader *arrayHeader = (ArrayHeader *)object;
+                size_t length = arrayHeader->length;
+                word_t **fields = (word_t **)(arrayHeader + 1);
+                if (length <= ARRAY_SPLIT_THRESHOLD) {
+                    Marker_markRange(heap, stats, in, outHolder, bytemap, fields, length);
+                } else {
+                    if (GreyPacket_IsEmpty(in)) {
+                        // last item - deal with it now
+
+                        // leave the last batch for the current thread
+                        word_t **limit = fields + length;
+                        word_t **lastBatch = fields + (length / ARRAY_SPLIT_BATCH) * ARRAY_SPLIT_BATCH;
+
+                        assert(lastBatch <= limit);
+                        for (word_t **batchFields = fields; batchFields < limit; batchFields += ARRAY_SPLIT_BATCH) {
+                            GreyPacket *slice = Marker_takeEmptyPacket(heap, stats);
+                            assert(slice != NULL);
+                            slice->type = grey_packet_refrange;
+                            slice->items[0] = (Stack_Type) batchFields;
+                            // no point writing the size, because it is constant
+                            Marker_giveFullPacket(heap, stats, slice);
+                        }
+
+                        size_t lastBatchSize = limit - lastBatch;
+                        if (lastBatchSize > 0) {
+                            Marker_markRange(heap, stats , in, outHolder, bytemap, lastBatch, lastBatchSize);
+                        }
+                    } else {
+                        // pass it on to someone else
+                        GreyPacket *slice = Marker_takeEmptyPacket(heap, stats);
+                        assert(slice != NULL);
+                        GreyPacket_Push(slice, object);
+                        Marker_giveFullPacket(heap, stats, slice);
+                    }
+                }
+            }
+            // non-object arrays do not contain pointers
+        } else {
+            int64_t *ptr_map = object->rtti->refMapStruct;
+            int i = 0;
```

Comment:
You can remove the need for this index via pointer arithmetics on a var that points to `ptr_map`, this is idiomatic in C:
```
int64_t *cursor = ptr_map;
for (*cursor != LAST_FIELD_OFFSET) {
   word_t *field = object->fields[*cursor];
   ...
   cursor++;
}
```